strategy/src/lib.rs

```diff
@@ -0,0 +1,407 @@
+//!     let future = FutureWrapper::new(wait_three_seconds());
+//!     future.await;
+//! });
+//! ```
+
+#![cfg_attr(not(feature = "std"), no_std)]
+#![cfg_attr(docsrs, feature(doc_cfg))]
+#![forbid(unsafe_code, future_incompatible, missing_docs)]
+
+use core::future::Future;
+use core::marker::PhantomData;
+use core::pin::Pin;
+use core::task::{Context, Poll};
+
+use event_listener::EventListener;
+use pin_utils::pin_mut;
+
+/// A wrapper around an [`EventListenerFuture`] that can be easily exported for use.
+///
+/// This type implements [`Future`], has a `_new()` constructor, and a `wait()` method
+/// that uses the [`Blocking`] strategy to poll the future until it is ready.
+///
+/// # Examples
+///
+/// ```
+/// mod my_future {
+///     use event_listener_strategy::{easy_wrapper, EventListenerFuture, Strategy};
+///     use std::pin::Pin;
+///     use std::task::Poll;
+///
+///     struct MyFuture;
+///
+///     impl EventListenerFuture for MyFuture {
+///         type Output = ();
+///
+///         fn poll_with_strategy<S: Strategy>(
+///             self: Pin<&mut Self>,
+///             strategy: &mut S,
+///             context: &mut S::Context,
+///         ) -> Poll<Self::Output> {
+///             /* ... */
+/// #           Poll::Ready(())
+///         }
+///     }
+///
+///     easy_wrapper! {
+///         /// A future that does something.
+///         pub struct MyFutureWrapper(MyFuture => ());
+///         /// Wait for it.
+///         pub wait();
+///     }
+///
+///     impl MyFutureWrapper {
+///         /// Create a new instance of the future.
+///         pub fn new() -> Self {
+///             Self::_new(MyFuture)
+///         }
+///     }
+/// }
+///
+/// use my_future::MyFutureWrapper;
+///
+/// // Use the future in a blocking context.
+/// let future = MyFutureWrapper::new();
+/// future.wait();
+///
+/// // Use the future in a non-blocking context.
+/// futures_lite::future::block_on(async {
+///     let future = MyFutureWrapper::new();
+///     future.await;
+/// });
+/// ```
+#[macro_export]
+macro_rules! easy_wrapper {
+    (
+        $(#[$meta:meta])*
+        $vis:vis struct $name:ident ($inner:ty => $output:ty);
+        $(#[$wait_meta:meta])*
+        $wait_vis: vis wait();
+    ) => {
+        ::pin_project_lite::pin_project! {
+            $(#[$meta])*
+            $vis struct $name {
+                #[pin]
+                _inner: $crate::FutureWrapper<$inner>
+            }
+        }
+
+        impl $name {
+            fn _new(inner: $inner) -> Self {
+                Self {
+                    _inner: $crate::FutureWrapper::new(inner)
+                }
+            }
+
+            $(#[$wait_meta])*
+            $wait_vis fn wait(self) -> $output {
+                use $crate::EventListenerFuture;
+                self._inner.into_inner().wait()
+            }
+        }
+
+        impl ::core::future::Future for $name {
+            type Output = $output;
+
+            fn poll(
```

Comment:
it might be a good idea to mark this with `#[inline]`

Reply:
I've always wondered if there is any emperical evidence of that helping anything. :thinking: Aren't the compilers these days smart enough to handle that, but with the added advantage of keeping the user-selected optimization level in mind?

Reply:
using `#[inline]` should just act as a hint, so it shouldn't massively hurt... I'm not how big the effect would be, benchmarking could sort-of prove it's effect...

Reply:
I know what they are intended for. They **do** hurt readability a bit so if there is no empirical evidence of them actually being useful, I would rather not have them.

Reply:
I asked the internets and the call was heeded: https://toot.cat/@jamey/110023643339747938 Reading the relevant part of that Rust book pointed there, and seeing it's a oneliner, I think you're right. This should be inlined across crates.

Sorry for the noise. I was mainly just wanting to make sure that extra line is worth it.

Reply:
Okay, I've inlined most of the small functions in this crate.